### PR TITLE
fix(Core/PetAI): Fixed logic in PetAI to determine if pet can cast co…

### LINKED ISSUE #4992: Warlock's imp doesn't auto cast Blood Pact [$10]
##### CURRENT BEHAVIOUR:
-With auto cast on, the imp won't auto cast Blood Pact even if it's not activated

##### EXPECTED BLIZZLIKE BEHAVIOUR:
If blood pact is not activated and auto cast is on, always activate it

##### SOURCE:
83e430f28a25b33175569749a0aec46ccc92c9b6

##### STEPS TO REPRODUCE THE PROBLEM:
1. Play warlock, summon an imp
2. Toggle auto cast (right click the ability)
3. try many different things to try to get him to auto cast the buff, like dying, resummoning etc

##### EXTRA NOTES:
Seems like he will auto cast it only if he gets atatcked, just like his other ability Fire shield, probably a bug related to that.

##### AC HASH/COMMIT:
83e430f28a25b33175569749a0aec46ccc92c9b6

##### CUSTOMIZATIONS:

<bountysource-plugin>

---
There is a **[$10 open bounty](https://www.bountysource.com/issues/97357179-warlock-s-imp-doesn-t-auto-cast-blood-pact?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
…mbat spells.

Fixed #4992.

## Changes Proposed:
-  Fixed wrong logic.
-  

## Issues Addressed:
- Closes  #4992

## Tests Performed:

-  Tested ingame with Imp.
- 

## How to Test the Changes:

1. Create warlock.
2. `.level 79`
3. `.learn all my class`
4. Summon Imp
5. Set Blood Pact to autocast spell

## Known Issues and TODO List:

- [ ]
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by UltraNix** [comment]
@T1ti  Please, test it.

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by pklloveyou** [comment]
Tested

**Comment by FrancescoBorzi** [comment]
@pklloveyou can you please be more specific about what have you tested?

for example, have you tested only the issue fixed by this PR or have you also tested pets in general to make sure there is no regression (i.e. no new bugs) ?

**Comment by pklloveyou** [comment]
I also tested hunter and mage pets

**Comment by FrancescoBorzi** [comment]
> I also tested hunter and mage pets

ok good, thank you

**Comment by pklloveyou** [comment]
@FrancescoBorzi We can't merge yet

**Comment by pklloveyou** [comment]
Not in combat, hunter pets are releasing skills all the time. Is that wrong?

**Comment by UltraNix** [comment]
If they are useable out of combat - why not? But I'm not sure - need confirmation.
@pklloveyou What specific skills do you mean?

**Comment by pklloveyou** [comment]
Pets are constantly releasing skills, but not in combat

**Comment by pklloveyou** [comment]
Well, maybe someone else should test it

**Comment by UltraNix** [comment]
Once again - what specific skills do you mean?

**Comment by pklloveyou** [comment]
image

File: src/server/game/AI/CoreAI/PetAI.cpp [hunk 1]
```diff
@@ -243,12 +243,9 @@ void PetAI::UpdateAI(uint32 diff)
...
             {
-                if (spellInfo->CanBeUsedInCombat())
-                {
-                    // Check if we're in combat or commanded to attack
-                    if (!me->IsInCombat() && !me->GetCharmInfo()->IsCommandAttack())
-                        continue;
-                }
+                // Check if we're in combat or commanded to attack
+                if ((me->IsInCombat() || me->GetCharmInfo()->IsCommandAttack()) && !spellInfo->CanBeUsedInCombat())
+                    continue;
```
Comment (Helias) [inline src/server/game/AI/CoreAI/PetAI.cpp:248]:
```suggestion
                if ((me->IsInCombat() || me->GetCharmInfo()->IsCommandAttack()) && !spellInfo->CanBeUsedInCombat())
                {
                    continue;
                }
```

**Review by Helias — CHANGES_REQUESTED** [review]

**Comment by FrancescoBorzi** [comment]
> @FrancescoBorzi We can't merge yet

ok, waiting for more feedback then

**Comment by UltraNix** [comment]
@pklloveyou Test again (Blood Pact too).

**Comment by pklloveyou** [comment]
Tested

**Pull request merged**